In [83]:
import pandas as pd
import numpy as np
import re

In [84]:
patients=pd.read_csv('patients.csv')
treatment=pd.read_csv('treatments.csv')
adverse=pd.read_csv('adverse_reactions.csv')
treatments_extra=pd.read_csv('treatments_cut.csv')

In [85]:
print(patients.shape)
print(treatment.shape)
print(adverse.shape)
print(treatments_extra.shape)

(503, 14)
(280, 7)
(34, 3)
(70, 7)


In [86]:
patients.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height   bmi  
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8  
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70  31.7  
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27  26.1

In [87]:
treatment.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [88]:
adverse.head()

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia
3     flavia  fiorentino                      cough
4    manouck     wubbels          throat irritation

In [89]:
treatments_extra.head()

given_name    surname    auralin    novodra  hba1c_start  hba1c_end  \
0      jožka  resanovič  22u - 30u          -         7.56       7.22   
1  inunnguaq   heilmann  57u - 67u          -         7.85       7.45   
2      alwin   svensson  36u - 39u          -         7.78       7.34   
3        thể      lương          -  61u - 64u         7.64       7.22   
4     amanda    ribeiro  36u - 44u          -         7.85       7.47   

   hba1c_change  
0          0.34  
1           NaN  
2           NaN  
3          0.92  
4          0.38

In [90]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [91]:
treatment.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [92]:
adverse.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
given_name          34 non-null object
surname             34 non-null object
adverse_reaction    34 non-null object
dtypes: object(3)
memory usage: 944.0+ bytes


In [93]:
treatments_extra.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70 entries, 0 to 69
Data columns (total 7 columns):
given_name      70 non-null object
surname         70 non-null object
auralin         70 non-null object
novodra         70 non-null object
hba1c_start     70 non-null float64
hba1c_end       70 non-null float64
hba1c_change    42 non-null float64
dtypes: float64(3), object(4)
memory usage: 4.0+ KB


In [94]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [95]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [96]:
treatment.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [97]:
treatments_extra.describe()

hba1c_start  hba1c_end  hba1c_change
count    70.000000  70.000000     42.000000
mean      7.838000   7.443143      0.518810
std       0.423007   0.418706      0.270719
min       7.510000   7.020000      0.280000
25%       7.640000   7.232500      0.340000
50%       7.730000   7.345000      0.370000
75%       7.860000   7.467500      0.907500
max       9.910000   9.460000      0.970000

In [98]:
patients[patients['address'].duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

### Quality

[Completeness, Validity, Accuracy, Consistency]

#### `Patients`

- Zip code col is stored in decimal format[V]
- Certain zip codes of 4 digits[V] 
- For the state col sometimes full names are use and other times short forms are used[IC]
- Missing values for contact details of 12 patients[C]
- Incorrect data types for assigned_sex and birthdate cols[?]
- Erroneous data for weights and heights[Accuracy]
- Inconsistent way of storing phone numbers in contact col[IC]
- Duplicate and not useful data in dataset[IC]
- birthdate col is in mm/dd/yyyy format [accuracy]
- the given_name and surname first letter is in caps compared to other tables[accuracy]

#### `Treatments`

- hba1c_change col has missing values[C]
- Novodra and Auralin cols has u attached to numbers[V]
- Only 280 rows are there, instead of 350[C]
- Name and surname col is in lower case[IC]
- Incorrect data type for auralin and novadra cols[V]
- hba1c_change 9 instead of 4[Accuracy]

#### `Adverse Reactions`

- Name and surname col is in lower case[IC]

#### Treatments extra

- has missing values in hba1c_change col[C]
- Incorrect datatype for auralin and novadra cols[V]

### Tidiness

#### `Patients`

- Phone and email together in one col i.e. contact

#### `Treatments`

- 3 variables stored in auralin and novodra cols
- use of '-' instead of nan in missing values

#### `Adverse Reactions`

- No point of using this table

#### Treatment extra 
- will merge this table with treatment


# Cleaning Data

You should follow the order provided below:

- Completeness Issues
- Tidiness Issues
- Remaining Data quality issues like validity, accuracy and consistency

### Steps involved in Data cleaning

- Define
- Code 
- Test

`Always make sure to create a copy of your pandas dataframe before you start the cleaning process`

In [99]:
# Creating copies

patients_copy=patients.copy()
treatment_copy=treatment.copy()
adverse_copy=adverse.copy()
treatments_extra_copy=treatments_extra.copy()

In [100]:
patients_copy.shape

(503, 14)

In [101]:
treatment_copy.shape

(280, 7)

In [102]:
adverse_copy.shape

(34, 3)

In [103]:
treatments_extra_copy.shape

(70, 7)

## a. Completeness Issues

- Missing values for contact details of 12 patients
- Only 280 rows are there, instead of 350
- hba1c_change col has missing values

#### Define : Missing values for contact details of 12 patients

##### solution : Nothing can be done to solve this problrm

#### Define : hba1c_change col has missing values

##### solution : Subtract hba1c_start from hba1c_end to get hba1c_change

#### Code

In [104]:
treatment_copy['hba1c_change']=treatment_copy['hba1c_start']-treatment_copy['hba1c_end']
treatments_extra_copy['hba1c_change']=treatments_extra_copy['hba1c_start']-treatments_extra_copy['hba1c_end']

#### Test

In [105]:
treatments_extra_copy.describe()

hba1c_start  hba1c_end  hba1c_change
count    70.000000  70.000000     70.000000
mean      7.838000   7.443143      0.394857
std       0.423007   0.418706      0.055995
min       7.510000   7.020000      0.280000
25%       7.640000   7.232500      0.350000
50%       7.730000   7.345000      0.390000
75%       7.860000   7.467500      0.440000
max       9.910000   9.460000      0.510000

In [106]:
treatment_copy.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    280.000000
mean      7.985929    7.589286      0.396643
std       0.568638    0.569672      0.060168
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.350000
50%       7.800000    7.420000      0.390000
75%       7.970000    7.570000      0.440000
max       9.950000    9.580000      0.540000

#### Define : Only 280 rows are there, instead of 350

#### Solution : joining treatment and treatment extra 

#### Code

In [107]:
treatment_copy=treatment_copy.append(treatments_extra_copy, ignore_index=True)

In [108]:
treatment_copy.shape

(350, 7)

#### Test

In [109]:
treatment_copy

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika    jindrová  41u - 48u          -         7.63       7.20   
1       elliot  richardson          -  40u - 45u         7.56       7.09   
2     yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3         skye  gormanston  33u - 36u          -         7.97       7.62   
4       alissa      montez          -  33u - 29u         7.78       7.46   
..         ...         ...        ...        ...          ...        ...   
345     rovzan     kishiev  32u - 37u          -         7.75       7.41   
346      jakob    jakobsen          -  28u - 26u         7.96       7.51   
347      bernd   schneider  48u - 56u          -         7.74       7.44   
348      berta  napolitani          -  42u - 44u         7.68       7.21   
349     armina       sauvé  36u - 46u          -         7.86       7.40   

     hba1c_change  
0            0.43  
1            0.47  
2            0.43  
3            0.35  
4            0.32  
..            ...  
345          0.34  
346          0.45  
347          0.30  
348          0.47  
349          0.46  

[350 rows x 7 columns]

## b. Tidiness Issues

- Phone + email in contact col `patients`
- Novodra and Aurolin col `treatment`
- Incorrect data type for dose 
- Merge the `adverse` table to `treatment` table


#### Define 

In [110]:
#Phone + email in contact col patients

#### Code

In [111]:
patients_copy['contact']=patients_copy['contact'].astype('str')
phone_no=[]

for ph in patients['contact']:
    try :
        var=re.findall('[0-9]+', ph)
        var='-'.join(var)[-12:]
        phone_no.append(var)        
    except:        
        phone_no.append(np.nan)        

In [112]:
patients_copy['email']=patients_copy.contact.str.extract(pat = '([a-z,A-Z,@,.]+)',expand=False)
patients_copy['phone_number']=phone_no
patients_copy.drop(columns='contact',inplace=True)

#### Test

In [113]:
patients_copy

patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liêm         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state  zip_code  \
0         576 Brown Bear Drive  Rancho California  California   92390.0   
1    2370 University Hill Road          Armstrong    Illinois   61812.0   
2        1493 Poling Farm Road               York    Nebraska   68467.0   
3          2335 Webster Street         Woodbridge          NJ    7095.0   
4         1428 Turkey Pen Lane             Dothan          AL   36303.0   
..                         ...                ...         ...       ...   
498        2530 Victoria Court       Milton Mills          ME    3852.0   
499   494 Clarksburg Park Road             Sedona          AZ   86341.0   
500          649 Nutter Street      Overland Park          MO   64110.0   
501   3652 Boone Crockett Lane            Seattle          WA   98109.0   
502          2778 North Avenue               Burr    Nebraska   68324.0   

           country  birthdate  weight  height   bmi  \
0    United States  7/10/1976   121.7      66  19.6   
1    United States   4/3/1967   118.8      66  19.2   
2    United States  2/19/1980   177.8      71  24.8   
3    United States  7/26/1951   220.9      70  31.7   
4    United States  2/18/1928   192.3      27  26.1   
..             ...        ...     ...     ...   ...   
498  United States  4/10/1959   181.1      72  24.6   
499  United States  3/26/1948   239.6      70  34.4   
500  United States  1/13/1971   171.2      67  26.8   
501  United States  2/13/1952   176.9      67  27.7   
502  United States   5/3/1954   138.2      71  19.3   

                                 email  phone_number  
0             ZoeWellish@superrito.com  951-719-9170  
1                 PamelaSHill@cuvox.de  217-569-3204  
2                 JaeMDebord@gustr.com  402-363-6804  
3            PhanBaLiem@jourrapide.com  732-636-8246  
4                  TimNeudorf@cuvox.de  334-515-7487  
..                                 ...           ...  
498    MustafaLindstrom@jourrapide.com  207-477-0579  
499             RumanBisliev@gustr.com  928-284-4492  
500          JinkedeKeizer@teleworm.us  816-223-6007  
501  ChidaluOnyekaozulu@jourrapide.com  360-443-2060  
502           PatrickGersten@rhyta.com  402-848-4923  

[503 rows x 15 columns]

#### Define

In [114]:
# Novodra and Aurolin using the melt function

#### Code

In [115]:
treatment_copy=pd.melt(treatment_copy, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'],
                           var_name='treatment', value_name='dose')

In [116]:
treatment_copy=treatment_copy[treatment_copy['dose']!='-']

In [117]:
treatment_copy['start'],treatment_copy['end']=treatment_copy['dose'].str.split('-').str

In [118]:
treatment_copy.drop(columns={'dose'}, inplace=True)

#### Test

In [119]:
treatment_copy.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   
7      eddie      archer         7.89       7.55          0.34   auralin   
9       asia     woźniak         7.76       7.37          0.39   auralin   

  start   end  
0  41u    48u  
3  33u    36u  
6  37u    42u  
7  31u    38u  
9  30u    36u

#### Define 

In [120]:
#: Incorrect data type fot dose

#### Code

In [121]:
treatment_copy['start']=treatment_copy['start'].str.replace('u','')

In [122]:
treatment_copy['end']=treatment_copy['end'].str.replace('u','')

In [123]:
treatment_copy['start']=treatment_copy['start'].astype('int')
treatment_copy['end']=treatment_copy['end'].astype('int')

In [124]:
treatment_copy.reset_index(drop=True,inplace=True)

#### Test

In [125]:
treatment_copy

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0       veronika    jindrová         7.63       7.20          0.43   auralin   
1           skye  gormanston         7.97       7.62          0.35   auralin   
2         sophia      haugen         7.65       7.27          0.38   auralin   
3          eddie      archer         7.89       7.55          0.34   auralin   
4           asia     woźniak         7.76       7.37          0.39   auralin   
..           ...         ...          ...        ...           ...       ...   
345  christopher    woodward         7.51       7.06          0.45   novodra   
346        maret    sultygov         7.67       7.30          0.37   novodra   
347        lixue       hsueh         9.21       8.80          0.41   novodra   
348        jakob    jakobsen         7.96       7.51          0.45   novodra   
349        berta  napolitani         7.68       7.21          0.47   novodra   

     start  end  
0       41   48  
1       33   36  
2       37   42  
3       31   38  
4       30   36  
..     ...  ...  
345     55   51  
346     26   23  
347     22   23  
348     28   26  
349     42   44  

[350 rows x 8 columns]

#### Define

In [126]:
# Merge the adverse table to treatment table

#### Code

In [127]:
treatment_copy.shape

(350, 8)

In [128]:
adverse_copy.shape

(34, 3)

In [129]:
treatment_copy=treatment_copy.merge(adverse_copy,how='outer',on=['given_name','surname'],)

#### Test

In [130]:
treatment_copy

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0       veronika    jindrová         7.63       7.20          0.43   auralin   
1           skye  gormanston         7.97       7.62          0.35   auralin   
2         sophia      haugen         7.65       7.27          0.38   auralin   
3          eddie      archer         7.89       7.55          0.34   auralin   
4           asia     woźniak         7.76       7.37          0.39   auralin   
..           ...         ...          ...        ...           ...       ...   
345  christopher    woodward         7.51       7.06          0.45   novodra   
346        maret    sultygov         7.67       7.30          0.37   novodra   
347        lixue       hsueh         9.21       8.80          0.41   novodra   
348        jakob    jakobsen         7.96       7.51          0.45   novodra   
349        berta  napolitani         7.68       7.21          0.47   novodra   

     start  end           adverse_reaction  
0       41   48                        NaN  
1       33   36                        NaN  
2       37   42                        NaN  
3       31   38                        NaN  
4       30   36                        NaN  
..     ...  ...                        ...  
345     55   51                     nausea  
346     26   23                        NaN  
347     22   23  injection site discomfort  
348     28   26               hypoglycemia  
349     42   44  injection site discomfort  

[350 rows x 9 columns]

In [131]:
treatment_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 349
Data columns (total 9 columns):
given_name          350 non-null object
surname             350 non-null object
hba1c_start         350 non-null float64
hba1c_end           350 non-null float64
hba1c_change        350 non-null float64
treatment           350 non-null object
start               350 non-null int32
end                 350 non-null int32
adverse_reaction    35 non-null object
dtypes: float64(3), int32(2), object(4)
memory usage: 24.6+ KB


##  Validity issue

- Zip code col is stored in decimal format[V]
- Certain zip codes of 4 digits[V]

#### Define

In [132]:
#Zip code col is stored in decimal format and some are of 4 digit which is not possible

#### Code

In [133]:
patients_copy['zip_code']=patients_copy['zip_code'].astype('str')

In [134]:
patients_copy['zip_code']=patients_copy['zip_code'].str.replace('.0','',regex=False).str.pad(width=5,side='left',fillchar='0')

#### Test

In [135]:
patients_copy['zip_code']

0      92390
1      61812
2      68467
3      07095
4      36303
       ...  
498    03852
499    86341
500    64110
501    98109
502    68324
Name: zip_code, Length: 503, dtype: object

##  Accuracy issue

- birthdate col is in mm/dd/yyyy format [accuracy]
- Erroneous data for weights and heights[Accuracy]
- the given_name and surname first letter is in caps compared to other tables[accuracy]

#### Define

In [136]:
#the given_name and surname first letter is in uppercase in patient compared to other tables

#### Code

In [137]:
patients_copy['surname']=[str(i).lower() for i in patients_copy['surname']]
patients_copy['given_name']=[str(i).lower() for i in patients_copy['given_name']]

#### Test

In [138]:
patients_copy[['given_name','surname']]

given_name      surname
0          zoe      wellish
1       pamela         hill
2          jae       debord
3         liêm         phan
4          tim      neudorf
..         ...          ...
498    mustafa    lindström
499      ruman      bisliev
500      jinke    de keizer
501    chidalu  onyekaozulu
502        pat      gersten

[503 rows x 2 columns]

#### Define

In [139]:
#Erroneous data for weights and heights

#### Code

In [140]:
patients_copy[patients_copy['height']<50]

patient_id assigned_sex given_name  surname               address    city  \
4           5         male        tim  neudorf  1428 Turkey Pen Lane  Dothan   

  state zip_code        country  birthdate  weight  height   bmi  \
4    AL    36303  United States  2/18/1928   192.3      27  26.1   

                 email  phone_number  
4  TimNeudorf@cuvox.de  334-515-7487

In [141]:
# this height may be corret but in case of mathematical usage this value will cause an error in calculation 
# thus we have to remove this row

In [142]:
patients_copy.drop(4,inplace=True)

#### Test

In [143]:
patients_copy.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        zoe  wellish       576 Brown Bear Drive   
1           2       female     pamela     hill  2370 University Hill Road   
2           3         male        jae   debord      1493 Poling Farm Road   
3           4         male       liêm     phan        2335 Webster Street   
5           6         male     rafael    costa         1140 Willis Avenue   

                city       state zip_code        country  birthdate  weight  \
0  Rancho California  California    92390  United States  7/10/1976   121.7   
1          Armstrong    Illinois    61812  United States   4/3/1967   118.8   
2               York    Nebraska    68467  United States  2/19/1980   177.8   
3         Woodbridge          NJ    07095  United States  7/26/1951   220.9   
5      Daytona Beach     Florida    32114  United States  8/31/1931   183.9   

   height   bmi                         email  phone_number  
0      66  19.6      ZoeWellish@superrito.com  951-719-9170  
1      66  19.2          PamelaSHill@cuvox.de  217-569-3204  
2      71  24.8          JaeMDebord@gustr.com  402-363-6804  
3      70  31.7     PhanBaLiem@jourrapide.com  732-636-8246  
5      70  26.4  RafaelCardosoCosta@gustr.com  386-334-5237

#### Define

In [144]:
#birthdate col is in mm/dd/yyyy format [accuracy]

#### Code

In [145]:
patients_copy.rename(columns={'birthdate':'birth_date(mm/dd/yyyy)'},inplace=True)

#### Test

In [146]:
patients_copy.columns

Index(['patient_id', 'assigned_sex', 'given_name', 'surname', 'address',
       'city', 'state', 'zip_code', 'country', 'birth_date(mm/dd/yyyy)',
       'weight', 'height', 'bmi', 'email', 'phone_number'],
      dtype='object')

### Inconsistency Issues

- Duplicate and not useful data in dataset
- Name and surname col is in lower case

#### Define

In [147]:
#For the state col sometimes full names are use and other times short forms are used

#### Code

In [148]:
k=patients_copy['state'].tolist()

In [149]:
# we have to replace 'California' with 'CA' ,
# we have to replace 'Illinois' with 'IL',
# we have to replace 'Nebraska' with 'NE',
# we have to replace 'Florida' with 'FL',

# rest are in sort form only

In [150]:
patients_copy['state'].replace({'California':'CA','Illinois':'IL','Nebraska':'NE','Florida':'FL'},inplace=True)

#### Test

In [151]:
patients_copy['state'].tolist()

['CA',
 'IL',
 'NE',
 'NJ',
 'FL',
 'NV',
 'CA',
 'MO',
 'New York',
 'MI',
 'TN',
 'VA',
 'AL',
 'OK',
 'VA',
 'CA',
 'GA',
 'MT',
 'CA',
 'MA',
 'NJ',
 'NY',
 'FL',
 'New York',
 'NM',
 'IL',
 'LA',
 'PA',
 'New York',
 'CO',
 'TN',
 'CA',
 'IL',
 'ME',
 'New York',
 'WI',
 'SD',
 'MA',
 'MA',
 'GA',
 'MN',
 'FL',
 'WY',
 'ME',
 'TN',
 'PA',
 'GA',
 'IL',
 'OH',
 'PA',
 'IL',
 'IA',
 'MA',
 'MA',
 'NV',
 'GA',
 'CA',
 'PA',
 'NC',
 'CA',
 'CA',
 'MA',
 'CA',
 'NY',
 'IN',
 'CA',
 'MA',
 'WI',
 'IL',
 'OH',
 'CT',
 'LA',
 'MA',
 'OK',
 'CA',
 'MN',
 'KY',
 'MA',
 'DE',
 'New York',
 'OK',
 'GA',
 'OH',
 'NY',
 'FL',
 'MD',
 'MN',
 'GA',
 'AZ',
 'New York',
 'TX',
 'NJ',
 'FL',
 'FL',
 'TX',
 'ME',
 'NE',
 'AK',
 'New York',
 'ND',
 'MO',
 'AL',
 'FL',
 'AZ',
 'NC',
 'IN',
 'MD',
 'KS',
 'MS',
 'TX',
 'CA',
 'VA',
 'TX',
 'New York',
 'CA',
 'KY',
 'LA',
 'LA',
 'TX',
 'CA',
 'New York',
 'MI',
 'FL',
 'MA',
 'OK',
 'MA',
 'WA',
 'IN',
 'NY',
 'SC',
 'WV',
 'PA',
 'SC',
 'New York',
 '

#### Define

In [152]:
#Duplicate and not useful data in dataset

#### Code

In [153]:
# In treatment dataframe
treatment_copy[treatment_copy.duplicated(subset=['given_name','surname'])]

given_name surname  hba1c_start  hba1c_end  hba1c_change treatment  start  \
6     joseph     day          7.7       7.19          0.51   auralin     29   

   end adverse_reaction  
6   36     hypoglycemia

In [154]:
# In patients dataframe
patients_copy[patients_copy.duplicated(subset=['given_name','surname'])]

patient_id assigned_sex given_name surname          address      city  \
229         230         male       john     doe  123 Main Street  New York   
237         238         male       john     doe  123 Main Street  New York   
244         245         male       john     doe  123 Main Street  New York   
251         252         male       john     doe  123 Main Street  New York   
277         278         male       john     doe  123 Main Street  New York   

    state zip_code        country birth_date(mm/dd/yyyy)  weight  height  \
229    NY    12345  United States               1/1/1975   180.0      72   
237    NY    12345  United States               1/1/1975   180.0      72   
244    NY    12345  United States               1/1/1975   180.0      72   
251    NY    12345  United States               1/1/1975   180.0      72   
277    NY    12345  United States               1/1/1975   180.0      72   

      bmi              email phone_number  
229  24.4  johndoe@email.com   1234567890  
237  24.4  johndoe@email.com   1234567890  
244  24.4  johndoe@email.com   1234567890  
251  24.4  johndoe@email.com   1234567890  
277  24.4  johndoe@email.com   1234567890

In [155]:
treatment_copy.drop_duplicates(inplace=True)

In [156]:
patients_copy.drop_duplicates(subset=['given_name','surname'],inplace=True)

#### Test

In [157]:
patients_copy.shape

(497, 15)

In [158]:
treatment_copy.shape

(349, 9)

# Cleaned data

In [159]:
patients_new=patients_copy.copy()

In [160]:
treatment_new=treatment_copy.copy()

### DataFrame

In [161]:
patients_new

patient_id assigned_sex given_name      surname  \
0             1       female        zoe      wellish   
1             2       female     pamela         hill   
2             3         male        jae       debord   
3             4         male       liêm         phan   
5             6         male     rafael        costa   
..          ...          ...        ...          ...   
498         499         male    mustafa    lindström   
499         500         male      ruman      bisliev   
500         501       female      jinke    de keizer   
501         502       female    chidalu  onyekaozulu   
502         503         male        pat      gersten   

                       address               city state zip_code  \
0         576 Brown Bear Drive  Rancho California    CA    92390   
1    2370 University Hill Road          Armstrong    IL    61812   
2        1493 Poling Farm Road               York    NE    68467   
3          2335 Webster Street         Woodbridge    NJ    07095   
5           1140 Willis Avenue      Daytona Beach    FL    32114   
..                         ...                ...   ...      ...   
498        2530 Victoria Court       Milton Mills    ME    03852   
499   494 Clarksburg Park Road             Sedona    AZ    86341   
500          649 Nutter Street      Overland Park    MO    64110   
501   3652 Boone Crockett Lane            Seattle    WA    98109   
502          2778 North Avenue               Burr    NE    68324   

           country birth_date(mm/dd/yyyy)  weight  height   bmi  \
0    United States              7/10/1976   121.7      66  19.6   
1    United States               4/3/1967   118.8      66  19.2   
2    United States              2/19/1980   177.8      71  24.8   
3    United States              7/26/1951   220.9      70  31.7   
5    United States              8/31/1931   183.9      70  26.4   
..             ...                    ...     ...     ...   ...   
498  United States              4/10/1959   181.1      72  24.6   
499  United States              3/26/1948   239.6      70  34.4   
500  United States              1/13/1971   171.2      67  26.8   
501  United States              2/13/1952   176.9      67  27.7   
502  United States               5/3/1954   138.2      71  19.3   

                                 email  phone_number  
0             ZoeWellish@superrito.com  951-719-9170  
1                 PamelaSHill@cuvox.de  217-569-3204  
2                 JaeMDebord@gustr.com  402-363-6804  
3            PhanBaLiem@jourrapide.com  732-636-8246  
5         RafaelCardosoCosta@gustr.com  386-334-5237  
..                                 ...           ...  
498    MustafaLindstrom@jourrapide.com  207-477-0579  
499             RumanBisliev@gustr.com  928-284-4492  
500          JinkedeKeizer@teleworm.us  816-223-6007  
501  ChidaluOnyekaozulu@jourrapide.com  360-443-2060  
502           PatrickGersten@rhyta.com  402-848-4923  

[497 rows x 15 columns]

In [162]:
treatment_new

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0       veronika    jindrová         7.63       7.20          0.43   auralin   
1           skye  gormanston         7.97       7.62          0.35   auralin   
2         sophia      haugen         7.65       7.27          0.38   auralin   
3          eddie      archer         7.89       7.55          0.34   auralin   
4           asia     woźniak         7.76       7.37          0.39   auralin   
..           ...         ...          ...        ...           ...       ...   
345  christopher    woodward         7.51       7.06          0.45   novodra   
346        maret    sultygov         7.67       7.30          0.37   novodra   
347        lixue       hsueh         9.21       8.80          0.41   novodra   
348        jakob    jakobsen         7.96       7.51          0.45   novodra   
349        berta  napolitani         7.68       7.21          0.47   novodra   

     start  end           adverse_reaction  
0       41   48                        NaN  
1       33   36                        NaN  
2       37   42                        NaN  
3       31   38                        NaN  
4       30   36                        NaN  
..     ...  ...                        ...  
345     55   51                     nausea  
346     26   23                        NaN  
347     22   23  injection site discomfort  
348     28   26               hypoglycemia  
349     42   44  injection site discomfort  

[349 rows x 9 columns]

### What is Assessing Data?

The data analysis process consists of the following steps

1. Asking Questions
2. Data Wrangling<br><br>
    a. Gathering Data
            i.   CSV files
            ii.  APIs
            iii. Web Scraping
            iv.  Databases
    b. <b>Assessing Data</b><br>
    c. Cleaning Data<br><br>
3. Exploratory Data Analysis
4. Drawing Conclusion
5. Comunicating Results
    

### Assessing Data

In this step, the data is to be understood more deeply. Before implementing methods to clean it, you will definitely need to have a better idea about what the data is about.

### Types of Unclean Data

There are 2 kinds of unclean data

1. <u>Dirty Data</u> (Data with Quality issues):<br>
    Dirty data, also known as low quality data. Low quality data has content issues.<br>
2. <u>Messy Data</u> (Data with tidiness issues)<br>
    Messy data, also known as untidy data. Untidy data has structural issues.

### 1. Dirty Data

When the data has following kinds of issues:

1. Duplicated data
2. Missing Data
3. Corrupt Data
4. Inaccurate Data

This kind of data is known as Dirty data

### 2. Messy Data

When the data has structural issue is it known as Messy or untidy data.

Tidy data has the following properties:

1. Each variable forms a column
2. Each observation forms a row
3. Each observational unit forms a table

If a data violates any one of the above it is known as untidy data

Example:

<img src="untidy.png">

### Types of Assessment

There are 2 types of assessment styles

1. Manual - Looking through the data manually in google sheets
2. Programmatic - By using pandas functions such as info(), describe() or sample()

### Steps in Assessment

There are 2 steps involved in Assessment

1. Discover
2. Document

### Data Quality Dimensions

1. Completeness Issues
2. Validity Issues
3. Accuracy Issues
4. Consistency Issues